src/ucp/wireup/wireup_ep.c

```diff
@@ -156,6 +156,9 @@ ucp_wireup_ep_pending_req_release(uct_pending_req_t *self, void *arg)
     if (proxy_req->send.proxy.req->func == ucp_wireup_msg_progress) {
         req = ucs_container_of(proxy_req->send.proxy.req, ucp_request_t,
                                send.uct);
+        ucs_assertv(req->send.wireup.type != UCP_WIREUP_MSG_ACK,
```

Comment:
1. why it cannot be ACK?
2. we already have assertion when adding to pending. do we really also need an assertion when removing from pending?

Reply:
> 1\. why it cannot be ACK?

we're sending `WIREUP_MSG/ACK` on UCT endpoint which was already REMOTE_CONNECTED (i.e. there is no WIREUP_EP wrapper)

> 2\. we already have assertion when adding to pending. do we really also need an assertion when removing from pending?

removed

Reply:
> we're sending `WIREUP_MSG/ACK` on UCT endpoint which was already REMOTE_CONNECTED (i.e. there is no WIREUP_EP wrapper)
> 

Is sending ACK going to break something from wireup_ep perspective?
We should make assertions only on incorrect usage that may break something like memory leak/segfault/etc (here, from the perspective or wireup_ep), and not make it harder to change wireup logic in the future...
For example the assert in line 647 is OK because it make sure the subsequent call to ucs_free() is not a double-freee

Reply:
> Is sending ACK going to break something from wireup_ep perspective?

it is just unexpected behavior
but also we don't expect that WIREUP_EP will live after ACK sent, otherwise - all protocol messages will be put into WIREUP_EP's queue instead of sending through UCT EP, so the application will hang.

Reply:
The usage of WIREUP_ACK for protocols is outside the scope of wireup_ep code: wireup_ep should not care, and not limit how it's being used. 
Even though the assertion itself is correct, it is breaking abstraction between wireup_ep and wireup protocol logic.

Reply:
I see your point, it makes sense.
removed this assertion